File: frontend/src/services/impl/TeamTaskService.ts
import BaseService from '../BaseService';
import type {
    TaskPriority,
    TaskStatus,
    TaskType,
    TeamTaskRequestDto,
    TeamTaskResponseDto
} from "../../entities/teamTask/teamTask.dto.ts";
import type {PaginationListResponseDto} from "../../entities/wrappers/wrapper.dto.ts";

// Базові параметри для фільтрації та пошуку
interface TaskQueryParams {
    page: number;
    size: number;
    search?: string;
    status?: TaskStatus;
    type?: TaskType;
    priority?: TaskPriority;
}

class TeamTaskService extends BaseService {
    constructor() {
        super('/team-tasks');
    }

    public createTask(roundId: number, data: TeamTaskRequestDto): Promise<TeamTaskResponseDto> {
        return this.post<TeamTaskResponseDto>(`/${roundId}`, data);
    }

    public updateTask(teamTaskId: number, data: TeamTaskRequestDto): Promise<TeamTaskResponseDto> {
        return this.put<TeamTaskResponseDto>(`/${teamTaskId}`, data);
    }

    public deleteTask(teamTaskId: number): Promise<void> {
        return this.delete<void>(`/${teamTaskId}`);
    }

    public updateTaskMeta(
        teamTaskId: number,
        meta: { status?: TaskStatus; priority?: TaskPriority; type?: TaskType }
    ): Promise<TeamTaskResponseDto> {
        return this.patch<TeamTaskResponseDto>(`/${teamTaskId}/update-meta`, null, { params: meta });
    }

    public updateAssignee(teamTaskId: number, assigneeId: number): Promise<TeamTaskResponseDto> {
        return this.patch<TeamTaskResponseDto>(`/${teamTaskId}/assign/${assigneeId}`);
    }

    public getTasksForMyTeamAndRound(
        roundId: number,
        params: TaskQueryParams
    ): Promise<PaginationListResponseDto<TeamTaskResponseDto>> {
        return this.get<PaginationListResponseDto<TeamTaskResponseDto>>(`/round/${roundId}/my`, { params });
    }

    public getMyTasks(params: TaskQueryParams): Promise<PaginationListResponseDto<TeamTaskResponseDto>> {
        return this.get<PaginationListResponseDto<TeamTaskResponseDto>>('/my', { params });
    }
}

export const teamTaskService = new TeamTaskService();

File: backend/src/main/java/com/project/backend/services/interfaces/TeamTaskService.java
package com.project.backend.services.interfaces;

import com.project.backend.models.TeamTask;
import com.project.backend.models.User;
import com.project.backend.models.constants.TaskPriority;
import com.project.backend.models.constants.TaskStatus;
import com.project.backend.models.constants.TaskType;
import org.springframework.data.domain.Page;

public interface TeamTaskService {

    TeamTask createTask(User creator, Long roundId, TeamTask teamTask);

    TeamTask updateTask(Long teamTaskId, TeamTask teamTask);

    void deleteTask(Long teamTaskId);

    TeamTask updateTaskMeta(Long teamTaskId, TaskStatus status, TaskPriority priority, TaskType type);

    TeamTask updateAssignee(Long teamTaskId, Long assigneeId);

    Page<TeamTask> findAllByTeamIdAndRoundId(Integer page, Integer size, String search, User user, Long roundId, TaskStatus status, TaskType type, TaskPriority priority);

    Page<TeamTask> findAllForUser(Integer page, Integer size, String search, User user, TaskStatus status, TaskType type, TaskPriority priority);

    TeamTask findById(Long teamTaskId);
}
